### PR TITLE
[JIT] metacompile isinstance checks

### PR DESCRIPTION
**Pull request by eellison**
This is a series of PRs that will allow us to support adding [padding to conv](https://github.com/pytorch/pytorch/pull/22484) and also reduce the friction of adding method overloads that was brought up in  https://github.com/pytorch/pytorch/pull/23266.

This PR only compiles one if branch if the condition is an isinstance check. This is consistent with what mypy does; it does not report errors if a branch can be determined statically to be unreachable.

```
def foo(x):
    # type: (int) -> int
    if isinstance(x, str):
        return x["1"]
    return x + 1

reveal_type(foo) # no error, shows int -> int
``` 

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23886 [JIT] add support for overloading functions
* **#23885 [JIT] metacompile isinstance checks**

Differential Revision: [D16697092](https://our.internmc.facebook.com/intern/diff/D16697092)

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -12803,6 +12803,35 @@ def test({arg_str}):
 
             FileCheck().check_not("prim::PythonOp").run(cu.test.graph)
 
+    def test_isinstance_metacompile(self):
```
Comment (driazati) [inline test/test_jit.py]: This should also check something where the value could be either type, i.e. a class attribute

```python
class M(torch.nn.Module):
	def __init__(self, x):
		self.x = x
	...

torch.jit.script(M(2))
torch.jit.script(M('hi'))
```

Comment (eellison) [reply test/test_jit.py]: `isinstance` is supposed to throw if it can't be statically resolved (don't have runtime support); that's an existing [limitation](https://github.com/pytorch/pytorch/blob/8fc349f7be84de74fbdadaca518a69af947d6739/torch/csrc/jit/script/compiler.cpp#L2181)

File: torch/csrc/jit/script/compiler.cpp [hunk 2]
```diff
@@ -1272,18 +1272,44 @@ struct to_ir {
...
   void emitIf(const If& stmt) {
-    // NOTE: emitIf checks on If stmt condition to see if the cond AST kind ==
-    // is/is not, for such cases we do meta programming and disable emitting the
+    // NOTE: emitIf checks on If stmt condition to see if the cond AST is
+    // a potential none check with is/is not, or an isinstance check.
+    // for such cases we do meta programming and disable emitting the
     // corresponding branches
     Expr cond = stmt.cond();
+    bool isinstance_call = isIsInstanceCall(cond);
+    bool potential_none_check = !isinstance_call && isPotentialNoneCheck(cond);
```
Comment (Chillee) [inline torch/csrc/jit/script/compiler.cpp]: nit: Is `isIsInstanceCall(cond)` ever true when `isPotentialNoneCheck(cond)` is true? Otherwise, seems like this conditional can just be `bool potential_none_check = isPotentialNoneCheck(cond)`.

File: torch/csrc/jit/script/compiler.cpp [hunk 3]
```diff
@@ -1272,18 +1272,44 @@ struct to_ir {
...
       Value* cond_value = emitCond(cond);
-      emitIfElseBlocks(cond_value, stmt);
-      return;
+      return emitIfElseBlocks(cond_value, stmt);
+    }
+
+    if (isinstance_call) {
+      auto is_instance_result = emitSugaredExpr(cond, 1);
+      auto ivalue = toIValue(is_instance_result->asValue(cond.range(), method));
+      TORCH_INTERNAL_ASSERT(ivalue); // no support for runtime checks
```
Comment (Chillee) [inline torch/csrc/jit/script/compiler.cpp]: nit: Shouldn't this be an user facing error? My understanding is that `TORCH_INTERNAL_ASSERT` is reserved for things that should never happen.

**Review by Chillee — APPROVED** [review]
Looks fairly straightforward and logic checks out.

File: torch/csrc/jit/script/compiler.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/script/compiler.cpp]: This is maybe over-optimization.  `isPotentialNoneCheck(cond)` is never true when `isIsInstanceCall(cond)` is true. So here I am only computing `isPotentialNoneCheck` when `isIsInstanceCall` is false.

If you think i should remove it i will

File: torch/csrc/jit/script/compiler.cpp (hunk 3 above)
Comment (eellison) [reply torch/csrc/jit/script/compiler.cpp]: This should never happen. If you look at implementation of `isinstance` it always [returns a constant](https://github.com/pytorch/pytorch/blob/8fc349f7be84de74fbdadaca518a69af947d6739/torch/csrc/jit/script/compiler.cpp#L2196)

File: test/test_jit.py (hunk 1 above)
Comment (driazati) [reply test/test_jit.py]: It is still statically resolved here, each module statically has a different type for that attribute name
